### PR TITLE
Modernize codebase

### LINKED ISSUE #12: strlen(): Passing null to parameter #1 ($string) of type string is deprecated
This one is easy to fix since the code almost does it right.

Stacktrace:
```text
ErrorException: strlen(): Passing null to parameter #1 ($string) of type string is deprecated in vendor/storyblok/richtext-resolver/src/Schema.php:48
Stack trace:
#0 [internal function]: {closure}(8192, 'strlen(): Passi...', '/Users/user/P...', 48)
#1 vendor/storyblok/richtext-resolver/src/Schema.php(48): strlen(NULL)
#2 vendor/storyblok/richtext-resolver/src/Resolver.php(111): Storyblok\RichtextRender\Schema->Storyblok\RichtextRender\{closure}(Array)
#3 vendor/storyblok/richtext-resolver/src/Resolver.php(48): Storyblok\RichtextRender\Resolver->getMatchingMark(Array)
```

And that line looks like so:
https://github.com/storyblok/storyblok-php-richtext-renderer/blob/a3d313f645553ed3dc50bd78200eac5938a10a2f/src/Schema.php#L48

Simply swap the conditions and it would work just fine:
```php
if ($anchor !== null && strlen($anchor) !== 0) {
```

_side note: an operator is significantly faster than a function call_

_another side note: this problem only applies to PHP 8_

### PR DESCRIPTION
**Pull request by uuf6429**
Major changes:
- improves testing mechanism with:
  - phpunit test config
  - enables code coverage
  - namespaced tests
  - fixes some test class/file names
  - fixes some assertions with switched arguments
  - merged some tests that relate to the same class
  - use data providers in case of tests for the same method with different i/o
- improves the package overall with:
  - mandatory composer package metadata (package description, type etc)
  - avoids exporting unnecessary things like tests (using .gitattributes)
  - simplifies autoloading (PSR-4 is supported since ages, alternatively we can switch to PSR-0 for best of both worlds)
  - avoids locking users to specific versions by gitignore'ing and not committing, composer.lock (this is typical for libraries)
- and changes in code:
  - makes the package potentially extendable by using protected instead of private methods
  - specifies method access in case of a particular class that missed them (**this might be a BC break if people used methods like `Resolver->renderNode`**)
  - fixes some nonsense 'bugs' such as returning null in a constructor, using array_push instead of operator etc
  - also fixes #12 
  - render class is used as an instance, so that it can be replaced if needs be (instead of static methods) (**this is also probably a BC break if people used the static methods or extended the class**)

Remaining things to consider:
- [x] Specify supported PHP versions in composer.json - this should be the baseline supported syntax
       https://github.com/uuf6429/storyblok-php-richtext-renderer/pull/3
- [x] Set up github 'action' pipeline to run tests
       https://github.com/uuf6429/storyblok-php-richtext-renderer/pull/1
- [x] ..and/or maybe post code coverage somewhere like [codecov.io](https://codecov.io/)
       https://github.com/uuf6429/storyblok-php-richtext-renderer/pull/1

----

_Edit: on a side-note, I'm pleased to report that even before any of my changes, code coverage stood at ~98%!_

### CODE REVIEW

**Comment by roberto-butti** [comment]
Hi everyone,
I just jumped into this repository.
I think this PR is great and improves the package in terms of adherence to standards and lays the foundation for introducing code quality tools such as static code analysis tools.

**Review by roberto-butti — APPROVED** [review]

**Comment by uuf6429** [comment]
Since there is renewed interest on this, I'll try to finish the remaining ticks above (I'll go for codecov for code coverage, which is free for OS projects, unless something else is preferred - which I can also set up - just let me know).

I'll also fix #14 here as well as a small bug/improvement fix for anchors which we found but didn't report.

**Comment by roberto-butti** [comment]
> Since there is renewed interest on this, I'll try to finish the remaining ticks above (I'll go for [codecov](https://about.codecov.io/) for code coverage, which is free for OS projects unless something else is preferred - which I can also set up - just let me know).
> 
> I'll also fix #14 here as well as a small bug/improvement fix for anchors which we found but didn't report.

Hi @uuf6429 , I plan to make some improvements to this package as I did with the PHP-client (adding phpstan and other stuff).
Your PR is excellent and goes in the right direction. To speed up the process, I'm asking you to create 1 PR for 1 issue, and for the code coverage, for now, it is enough to have a script in composer that launch the tests with coverage support. This is useful for local development, and we will see shortly how to use this information in the CICD (for example, raising an error if the coverage doesn't reach a specific percentage of coverage)

**Comment by uuf6429** [comment]
> > Since there is renewed interest on this, I'll try to finish the remaining ticks above (I'll go for [codecov](https://about.codecov.io/) for code coverage, which is free for OS projects unless something else is preferred - which I can also set up - just let me know).
> > I'll also fix #14 here as well as a small bug/improvement fix for anchors which we found but didn't report.
> 
> Hi @uuf6429 , I plan to make some improvements to this package as I did with the PHP-client (adding phpstan and other stuff). Your PR is excellent and goes in the right direction. To speed up the process, I'm asking you to create 1 PR for 1 issue, and for the code coverage, for now, it is enough to have a script in composer that launch the tests with coverage support. This is useful for local development, and we will see shortly how to use this information in the CICD (for example, raising an error if the coverage doesn't reach a specific percentage of coverage)

Hi!

For the problems I mentioned, I'll make separate PRs as advised.
Question: should the fixes target this PR or also the old(current) codebase?

For code coverage / cicd, I already have some github workflows (in other projects) that run tests and also upload results to codecov. I can make a separate PR so you can give that a look (unless you're convinced not to go this way).

**Review by ademarCardoso — APPROVED** [review]
Seems to be working well and has been well refactored, good work friend 🚀 

Screenshot 2023-01-09 at 15 02 00

**Comment by uuf6429** [comment]
Hey, I just realised there's another PR with bigger changes: 💥 https://github.com/uuf6429/storyblok-php-richtext-renderer/pull/3

In my opinion, that final PR is the way to go, (it introduces interfaces, for example), but I understand it's a lot of work to get there (and perhaps going back and forward). If you guys want to speed this up and do it (review, update, release) in a sort of live hackathon / call, let me know.

**Review by joaokamun — APPROVED** [review]

**Comment by roberto-butti** [comment]
Hi @uuf6429 , thank you for the PR ✨ 
I'm going to merge it into the master branch.
Then I will fix a bit the composer .json for:
- requiring PHP from 7.4 to 8.2 (yes , in version 2.X, we have to support also PHP 7 for older software that requires the package)
- supporting phpunit8 and 9

All these changes will be tagged as 2.1.0

Then, we can "refactor" the whole package to achieve version 3.0.0. For future version 3, we will create a `develop` branch.

My colleague and friend @joaokamun  is aligned 🚀

**Comment by uuf6429** [comment]
@roberto-butti hey, I've been thinking..

I wouldn't put the PHP version requirement in v2.1, since that is a breaking change (right now there is no requirement at all, theoretically we're saying this packages works with PHP 5.4 😄). You could use a sane value like PHP 5.6, but I think that is a bit pointless and too late now.

I would however merge https://github.com/storyblok/storyblok-php-richtext-renderer/pull/15 as part of v2.1.

In my opinion, changing the PHPUnit version for v2.1 is also not beneficial. Keep in mind that both of the changes you mentioned are also in https://github.com/uuf6429/storyblok-php-richtext-renderer/pull/3 (which make sense for a v3).

So in short, my suggested roadmap would be (and almost no code changes on your side):
- merge https://github.com/storyblok/storyblok-php-richtext-renderer/pull/15
- (maybe) merge https://github.com/uuf6429/storyblok-php-richtext-renderer/pull/1
- tag v2.1
- merge https://github.com/uuf6429/storyblok-php-richtext-renderer/pull/3
- tag v3.0

**Comment by roberto-butti** [comment]
> @roberto-butti hey, I've been thinking..
> 
> I wouldn't put the PHP version requirement in v2.1, since that is a breaking change (right now there is no requirement at all, theoretically we're saying this packages works with PHP 5.4 😄). You could use a sane value like PHP 5.6, but I think that is a bit pointless and too late now.
> 
> I would however merge #15 as part of v2.1.
> 
> In my opinion, changing the PHPUnit version for v2.1 is also not beneficial. Keep in mind that both of the changes you mentioned are also in [uuf6429#3](https://github.com/uuf6429/storyblok-php-richtext-renderer/pull/3) (which make sense for a v3).
> 
> So in short, my suggested roadmap would be (and almost no code changes on your side):
> 
> * merge [Fix storyblok-php-richtext-renderer#14 #15](https://github.com/storyblok/storyblok-php-richtext-renderer/pull/15)
> * (maybe) merge [Add GitHub workflow; upgrade phpunit #16](https://github.com/storyblok/storyblok-php-richtext-renderer/pull/16)
> * tag v2.1
> * merge [Feature/improve typing uuf6429/storyblok-php-richtext-renderer#3](https://github.com/uuf6429/storyblok-php-richtext-renderer/pull/3)
> * tag v3.0

Hi @uuf6429 , first of al, thank you for your great feedback ,and sharing thoughts.
I try to reply:
- i agree to merge the #14, i added a comment for adding test. Feel free to let me know if you can add the test or i can do it
- i already merged a workflow
- i added ^8|^9 for phpunit allows composer to install the right dependency according with the php version. For example if you want to launch coverage with php8.2 you need phpunit9 because of the dependencies
- if someone has php5.6 can install 2.0 or 1.X
- because version 3 can takes time, I would like to have a kind of "parallel" release:
    - master for merging stuff for 2.1.x
    - develop for merging stuff for 3.0.0 (i want to include also PHPstan, we will define the level, and i want to include pint for PSR12)
    - once develop will be ready we can open a PR from develop to master

**Comment by uuf6429** [comment]
ok, sounds good, just one thing:

> - if someone has php5.6 can install 2.0 or 1.X

Just note that this is not usually how it works - people typically use version constraints like `^2.0`  in `composer.json` because BC breaks _should_ be reserved for major changes.
I _don't think_ this particular situation will cause any trouble - worst case they would have to change the constraint manually.
